### PR TITLE
Fix ModuleInfo skip logic

### LINKED ISSUE #80247: `ModuleInfo` skip logic is incorrect
The skip logic in `ModuleInfo` is based on the old version of `OpInfo`'s skip logic:

https://github.com/pytorch/pytorch/blob/9d6e3f81d46382ed91b6ff286efe07e5cf5dd052/torch/testing/_internal/common_modules.py#L198-L199

https://github.com/pytorch/pytorch/blob/9d6e3f81d46382ed91b6ff286efe07e5cf5dd052/torch/testing/_internal/common_modules.py#L107-L119

This logic is misleading and can lead to overly eager skipping. For example, consider this:

https://github.com/pytorch/pytorch/blob/9d6e3f81d46382ed91b6ff286efe07e5cf5dd052/torch/testing/_internal/common_modules.py#L1153-L1155

This `DecorateInfo` will be considered "active" for the test `TestModule.test_memory_format` and will result in *unconditional* skipping (notably, it won't check whether CUDA is present or the cuDNN version, even though intuitively this would be expected). The definition of "active" is here:

https://github.com/pytorch/pytorch/blob/bda04e9f5efc1dd8788a7172a2f8114430b1e1c4/torch/testing/_internal/common_methods_invocations.py#L99-L106

OpInfo's skip logic was refactored in #62713, fixing this problem. `ModuleInfo` should get the same refactoring to be more correct.

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #79979
* __->__ #80471

Fixes #80247

This PR:
* Refactors the skip logic as done for OpInfo in #62713, fixing the logic error
* For tests that were wrongly skipped before and now fail:
    * Fix `TestModule.test_cpu_gpu_parity` to support Lazy modules - this was affecting `LazyConv*`
    * Adds `@expectedFailure` decorators and a follow-up message to address `Conv*` failures on `TestModule.test_memory_format`

### CODE REVIEW

**Comment by jbschlosser** [comment]
@mruberry this is ready for review when you get back and have time :)

File: torch/testing/_internal/common_modules.py [hunk 1]
```diff
@@ -1227,6 +1243,11 @@ def module_inputs_torch_nn_LSTM(module_info, device, dtype, requires_grad, train
                    # See https://github.com/pytorch/pytorch/issues/70505 for more info.
                    DecorateInfo(skipMeta),
                    DecorateInfo(skipIfMps, 'TestModule', dtypes=[torch.float64]),
+                   # This was wrongly being skipped before and needs investigation.
```
Comment (mruberry) [inline torch/testing/_internal/common_modules.py:1246]: Nice comments

**Review by mruberry — APPROVED** [review]
Stamped; I'm assuming you spot-checked the test logs to ensure tests were running as expected (one of the gaps in our CI is the ability -- or inability -- to know if we're running the tests we expect to)

**Comment by jbschlosser** [comment]
> Stamped; I'm assuming you spot-checked the test logs to ensure tests were running as expected (one of the gaps in our CI is the ability -- or inability -- to know if we're running the tests we expect to)

Yep, good call making sure on this - I've spot checked and things are as expected from what I can tell.

**Pull request closed without merging**